include/envoy/stream_info/filter_state.h

```diff
@@ -50,6 +57,23 @@ class FilterState {
     return *result;
   }
 
+  /**
+   * @param data_name the name of the data being looked up.
+   * @return a non-const reference to the stored data if and only if the
+   * underlying data is mutable.
+   * An exception will be thrown if the data does not exist or if it is
+   * immutable. This function will fail if the data stored under
+   * |data_name| cannot be dynamically cast to the type specified.
+   */
+  template <typename T> T& getData(absl::string_view data_name) {
```

Comment:
Call to non-const ref/pointer of FilterState will be go here regardless of the data_name (and an exception will be thrown for immutable data), so this should be named differently than the const one above, name this `mutableData`?

Reply:
So the LHS doesn't matter at all? 
const auto& foo = filterState.getData() ?

Reply:
No, unless you cast filterState to const. i.e. `static_cast<const FilterState&>(filterState).getData()`